Implement xbmcvfs module

michaelarnauts opened the pull request:
### Functional description
<!--- Give a clear explanation of the change, fix or new feature 
that this PR contains -->
<!--- Put your text below this line -->
This PR implements the `xbmcvfs` class.

It basicly is a copy from https://github.com/add-ons/plugin.video.vrt.nu/blob/master/tests/xbmcvfs.py file from the VRT NU addon. I believe it was written by @dagwieers (credit where credit is due). 

Since Kodi Matrix, the functions `makeLegalFilename` and `validatePath` have been moved to `xbmcvfs`. I've added a implementation for them.

I still need to do some more testing if everything works fine, hence the draft status.

<!--- Put your text above this line -->

### Reasoning
<!--- Provide a decent justification for this PR -->
<!--- Put your text below this line -->

<!--- Put your text above this line -->

### Technical description
<!--- Please provide a technical analysis of this PR, explaining the 
 changes that were made. -->
<!--- Put your text below this line -->

<!--- Put your text above this line -->

basrieter commented on xbmcvfs.py:
```diff
@@ -0,0 +1,176 @@
+        def __init__(self, path):
+            """ Stat class.
+
+            :param path:    The file or directory to stat
```
please add types to the documentation, e.g.:

`:param str path:    The file or directory to stat`

This goes for all methods. IDE's use it for type checking.

basrieter commented on xbmcvfs.py:
```diff
@@ -0,0 +1,176 @@
+    return open(path, flags, encoding='utf-8')
+
+
+def Stat(path):  # NOSONAR
```
Why this weird construction? It should just be a class: 

```python
class State(object):
    def __init__(path):
        """" Initializes a Stat class.

        :param str path:    The file or directory to stat

        """"

        self._stat = os.stat(path)
```

michaelarnauts replied:
I'm not sure, this was like this in the code. I assumed that there needed to be a `Stat()` function or something. Maybe @dagwieers can tell more.

basrieter commented on xbmcvfs.py:
```diff
@@ -0,0 +1,176 @@
+            self._stat = os.stat(path)
+
+        def st_atime(self):
+            """ Returns the st_mtime attribute """
```
Please include return type and return parameters in all methods:

```python
        def st_atime(self):
            """ Returns the st_mtime attribute 
            
            :rtype: float
            :return: The st_mtime attribute
            
            """
            return self._stat.st_atime
```

basrieter commented on xbmcvfs.py:
```diff
@@ -0,0 +1,176 @@
+
+
+def File(path, flags='r'):  # NOSONAR
+    return open(path, flags, encoding='utf-8')
```
Please use `io.open()` to improve cross platform and cross Python version compatibility.  

I also don't know if this will work, because a Python file object is returned, not a Kodi File object. The biggest difference is the missing `readBytes`. But for now, let's do it like this.

michaelarnauts replied:
I've written out the functions for the `File` class. I did notice that it's also possible to use it with the `with` syntax from Kodi 19:

```
with xbmcvfs.File(file, 'w') as f:
  result = f.write(buffer)
```

I need to look into this, but I think I need to implement a `contextmanager` for this, altough I'm not sure if this also works with classes.

michaelarnauts replied:
This was easy :)

michaelarnauts commented:
Thanks for the review! I'll look at the remarks later.

Out of curiousity, what IDE do you use? I use PyCharm, but it seems that this type of documentation isn't that well implemented there.

basrieter commented:
I also use PyCharm (Pro), and as far as I can see, they work OK. One of the main reasons I do that. The only thing is that if you pass around custom objects, you need to `import` them in the module that has the documentation.

michaelarnauts commented:
I've rebased against `master` and added tests for the `File` and `Stat` classes. They should probably be ran in CI together with `pylint` or something.

michaelarnauts commented:
I see I'm still missing some types. I will add them later.

michaelarnauts commented:
@basrieter it might make sense to squash and merge instead of just merging, I don't think it matters to have all my in-between-commits in the master branch.

michaelarnauts commented:
I've rebased again to `master` to fix the conflict in `__init__.py`.

basrieter commented on xbmcvfs.py:
```diff
@@ -0,0 +1,286 @@
+        :param str path:    The file or directory to open
+        :param str flags:   The flags used to open the file
+        """
+        self._file = io.open(path, flags)
```
The flags from `def __init__(self, path, flags='r'):` are not the ones to be passed to the `io.open()`. Kodi accepts `r` or `w` and files are opened in binary mode. That is why there is a `read()` and `readBytes()`. The first does a `self._file.read().decode('utf-8')` the latter just a `self._file.read()`

basrieter commented on xbmcvfs.py:
```diff
@@ -0,0 +1,286 @@
+        :return: The read bytes
+        :rtype: str
+        """
+        return self._file.read(bytes)
```
should be `return self._file.read(bytes).decode('utf-8')`

basrieter commented on xbmcvfs.py:
```diff
@@ -0,0 +1,286 @@
+        """
+        return self._file.read(numbytes)
+
+    def seek(self, seekBytes, iWhence=0):  # NOSONAR
```
Should return an integer (See https://romanvm.github.io/Kodistubs/_modules/xbmcvfs.html#File.seek)

michaelarnauts replied:
I see, it seems that this is missing in the Kodi documentation.
-> https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/legacy/File.h#L219

michaelarnauts commented:
I'll try to do these changes in the course of this week.

michaelarnauts commented:
I've fixed your remarks, and also added a few tests. My `makedirs` command was also broken. It should also return `True` if the folder already exists.

michaelarnauts commented:
And I've missed that files should always be opened in binary mode. This is also corrected.

michaelarnauts commented:
I still get this error on Python 3:

```py
Traceback (most recent call last):
  File "/home/runner/work/plugin.video.vtm.go/plugin.video.vtm.go/tests/test_routing.py", line 52, in test_catalog_program_menu
    plugin.run([routing.url_for(plugin.show_catalog_program, program='e892cf10-5100-42ce-8d59-6b5c03cc2b96'), '0', ''])
  File "/home/runner/work/plugin.video.vtm.go/plugin.video.vtm.go/resources/lib/plugin.py", line 218, in run
    routing.run(params)
  File "/opt/hostedtoolcache/Python/3.5.9/x64/lib/python3.5/site-packages/routing.py", line 130, in run
    self._dispatch(self.path)
  File "/opt/hostedtoolcache/Python/3.5.9/x64/lib/python3.5/site-packages/routing.py", line 141, in _dispatch
    view_func(**kwargs)
  File "/home/runner/work/plugin.video.vtm.go/plugin.video.vtm.go/resources/lib/plugin.py", line 107, in show_catalog_program
    Catalog(kodi).show_program(program)
  File "/home/runner/work/plugin.video.vtm.go/plugin.video.vtm.go/resources/lib/modules/catalog.py", line 105, in show_program
    program_obj = self._vtm_go.get_program(program, cache=CACHE_PREVENT)  # Use CACHE_PREVENT since we want fresh data
  File "/home/runner/work/plugin.video.vtm.go/plugin.video.vtm.go/resources/lib/vtmgo/vtmgo.py", line 583, in get_program
    self._kodi.set_cache(['program', program_id], program)
  File "/home/runner/work/plugin.video.vtm.go/plugin.video.vtm.go/resources/lib/kodiwrapper.py", line 409, in set_cache
    json.dump(data, fdesc)
  File "/opt/hostedtoolcache/Python/3.5.9/x64/lib/python3.5/json/__init__.py", line 179, in dump
    fp.write(chunk)
  File "/opt/hostedtoolcache/Python/3.5.9/x64/lib/python3.5/site-packages/xbmcvfs.py", line 81, in write
    return self._file.write(buffer) == len(buffer)
TypeError: a bytes-like object is required, not 'str'
```
